src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.cs

```diff
@@ -157,12 +158,11 @@ protected override ImmutableArray<AbstractFormattingRule> GetCustomFormattingRul
         protected override SyntaxToken GetMethodNameAtInvocation(IEnumerable<SyntaxNodeOrToken> methodNames)
             => (SyntaxToken)methodNames.FirstOrDefault(t => !t.Parent.IsKind(SyntaxKind.MethodDeclaration));
 
-        protected override async Task<OperationStatus> CheckTypeAsync(
-            Document document,
+        protected override OperationStatus CheckType(
+            SemanticModel semanticModel,
             SyntaxNode contextNode,
             Location location,
-            ITypeSymbol type,
-            CancellationToken cancellationToken)
+            ITypeSymbol type)
```

Comment:
all these semantic checking helpers used to pass in the document, and then get the semantic model from that.  very bad.  each time we'd make it and throw it away.  we now just pass the semantic model in, and can do things synchronously on one model.

Reply:
For my understanding, this would use a cached semantic model if it had already been calculated for the document, right? The problem is if the document is forked and then this was being called?

Reply:
No. We don't cache semantic models. We hold onto them weakly. So you'll get the same one back if anyone else is holding onto it. But if no one is, it will make a new instance.  

In the non-forking case it's less bad (since usually someone is holding onto the model somewhere). In the forking case it's very bad since no one is.